src/dune_console/dune_console.ml

```diff
@@ -189,33 +241,56 @@ module Backend = struct
       let open T in
       let last = ref (Unix.gettimeofday ()) in
       let frame_rate = 1. /. 60. in
+      let cleanup () =
+        state.finished <- true;
+        Base.finish ();
+        Condition.broadcast finish_cv;
+        Mutex.unlock mutex
+      in
       try
+        Base.start ();
         while true do
           Mutex.lock mutex;
-          while not (Queue.is_empty state.messages) do
-            Base.print_user_message (Queue.pop_exn state.messages)
-          done;
-          Base.set_status_line state.status_line;
-          flush stderr;
+          Base.render state;
           let finish_requested = state.finish_requested in
           if finish_requested then raise_notrace Exit;
           Mutex.unlock mutex;
           let now = Unix.gettimeofday () in
           let elapsed = now -. !last in
-          if elapsed >= frame_rate then last := now
-          else
-            let delta = frame_rate -. elapsed in
-            Unix.sleepf delta;
-            last := delta +. now
+          let new_time =
+            if elapsed >= frame_rate then
+              Base.handle_user_events ~now ~timeout:0.0 mutex
+            else
+              let delta = frame_rate -. elapsed in
+              Base.handle_user_events ~now ~timeout:delta mutex
+          in
+          last := new_time
         done
-      with Exit ->
-        state.finished <- true;
-        Condition.broadcast finish_cv;
-        Mutex.unlock mutex );
+      with
+      | Exit -> cleanup ()
+      | exn ->
+        let exn = Exn_with_backtrace.capture exn in
+        cleanup ();
+        Exn_with_backtrace.reraise exn );
     (module T)
 
-  let progress =
-    let t = lazy (threaded (module Progress_no_flush)) in
+  module Progress_no_flush_threaded : Threaded = struct
+    include Progress_no_flush
+
+    let render state =
+      while not (Queue.is_empty state.messages) do
+        print_user_message (Queue.pop_exn state.messages)
+      done;
+      set_status_line state.status_line;
+      flush stderr
+
+    let handle_user_events ~now ~timeout _ =
```

Comment:
The current console doesn't react to user events so we just sleep until the next loop iteration. Because it doesn't react to user input, it cannot modify the ui state, and as a consequence doesn't need the mutex

Reply:
Why not add this comment to the code?